rfc/006-semantic-versioning.md

```diff
@@ -0,0 +1,76 @@
+  
+4. There is also a fourth emerging user: IBC module and relayers (although in many ways relayers can also be considered as external clients). The developers of IBC and consumers of their software are concerned about changes that may affect a chains ability to send arbitrary messages to another chain. Specifically, this can happen with a breaking change to the light client verification algorithm.  
+
+In a more broader sense, these four users are inextricably linked and their concerns reflect that of the end user of the applications. The major conclusion that this RFC makes is that **the ability for chains to provide continual service is more important than the actual upgrade burden put on the developers of these chains**. What that means is that although it may be burdensome for application developers to upgrade Tendermint versions, this is far outweighed by the pain node operators experience in upgrading, by the pain that external clients have in accommodating changes and ultimately the pain of end users who are benefiting from the application. Developers have all the time in world to make these changes whilst for other user groups, upgrading can be considered part of this real-time critical path.  
+
+## Proposal
+
+Tendermint will follow a same formatting structure of [SemVer](https://semver.org/) with a major, minor and patch version but with slight adjustments to what they encompass:
+
+A major version in Tendermint guarantees that for the entire cycle of that major version:
+
+- All blocks and state data in a blockchain can be queried. All headers can be verified even across minor version changes. Nodes can both block sync and state sync from genesis to the head of the chain.
+- Nodes in a network are able to communicate and perform BFT state machine replication so long as the agreed network version is the lowest of all nodes in a network. For example, nodes using version 1.5.x and 1.2.x can operate together so long as the network version is 1.2 or lower (but still within the 1.x range). This rule essentially captures the concept of network backwards compatibility.
+- An external client should not need to modify their code to interact with a node's RPC endpoints. New endpoints can be added but existing ones can not be altered or removed.
+- Migrations should be automatic. Upgrading of one node can happen asynchronously with respect to other nodes (although agreement of a network-wide upgrade will happen synchronously via consensus).
+
+A minor version in Tendermint guarantees that for the entire cycle of that minor version:
+
+- All publicly available Go API's, for example in `node` or `abci` packages will not change in a breaking way that would require any consumer (not just application developers) to modify their code.
+- No breaking changes to the block protocol. This means that all block related data structures should not change in a way that breaks any of the hashes, the consensus engine or light client verification.
```

Comment:
`hash-breaking` -> `verification-breaking`?

Reply:
> `hash-breaking` ⇒ `verification-breaking`?

That seems fine too—but I'm not sure it is any clearer? Isn't verification essentially an extension of hash equality?

Reply:
For the most part, although if we change the signature scheme, say to aggregated signatures, we haven't really broken the hash verification but we have produced a 'verification' breaking change since the old code can no longer verify the new block structure.

Reply:
This is true. I think verification breaking is more accurate although I'll keep the examples of changing the contents in a hash or changes in signature verification

Reply:
I see, so the issue is not just changes to the block _structure_, but also changes to the protocol rules (e.g., your example of the hash algorithm). That makes sense.